pmd-core/src/main/java/net/sourceforge/pmd/lang/document/TextDocument.java

```diff
@@ -0,0 +1,255 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.document;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.Reader;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+import net.sourceforge.pmd.cpd.SourceCode;
+import net.sourceforge.pmd.lang.LanguageVersion;
+import net.sourceforge.pmd.util.datasource.DataSource;
+
+/**
+ * Represents a textual document, providing methods to edit it incrementally
+ * and address regions of text. A text document delegates IO operations
+ * to a {@link TextFile}. It reflects some in-memory snapshot of the file,
+ * though the file may still be edited externally.
+ *
+ * <p>TextDocument is meant to replace CPD's {@link SourceCode} and PMD's
+ * {@link DataSource}, though the abstraction level of {@link DataSource}
+ * is the {@link TextFile}.
+ *
+ * <p>Note that the backing {@link TextFile} is purposefully not accessible
+ * from a text document. Exposing it here could lead to files being written
+ * to from within rules, while we want to eventually build an API that allows
+ * file edition based on AST manipulation.
+ *
+ * <h3>Coordinates in TextDocument</h3>
+ *
+ * This interface is an abstraction over a piece of text, which might not
+ * correspond to the backing source file. This allows the document to
+ * be a view on a piece of a larger document (eg, a Javadoc comment, or
+ * a string in which a language is injected). Another use case is to perform
+ * escape translation, while preserving the line breaks of the original source.
```

Comment:
The translation of escape codes that happens before lexing in some languages. Eg in java these are the unicode escapes like `\u00a0`, in C++ the line continuations (`\` + newline). In the followup branch, they are decoded in a single pass over the original text document. That pass produces a new text document with the translated source (escapes replaced with the characters they stand for), which is then passed to the lexer/parser. Doing this in a separate pass simplifies the CharStream implementation for Javacc, and also makes sure that the text that is available in rules doesn't contain escapes. Otherwise you could eg  have a `class` token whose image is `\u0064lass`, so the `getText` method would be less useful